apps/survey/contracts/Survey.sol

```diff
@@ -98,84 +99,104 @@ contract Survey is AragonApp {
+     * @param _surveyId Id for survey
+     */
+    function resetVote(uint256 _surveyId) public {
+        require(canVote(_surveyId, msg.sender));
+
+        SurveyStruct storage survey = surveys[_surveyId];
+        MultiOptionVote storage previousVote = survey.votes[msg.sender];
+        if (previousVote.optionsCastedLength == 0) {
+            return;
+        }
+
+        // voter removes their vote
+        for (uint256 i = 1; i <= previousVote.optionsCastedLength; i++) {
+            OptionCast storage previousOptionCast = previousVote.castedVotes[i];
+            survey.optionVotes[previousOptionCast.optionId] = survey.optionVotes[previousOptionCast.optionId].sub(previousOptionCast.stake);
+        }
+
+        // compute previously casted votes (i.e. substract non-used tokens from stake)
+        uint256 voterStake = token.balanceOfAt(msg.sender, survey.snapshotBlock);
+        uint256 previousParticipation = voterStake.sub(previousVote.castedVotes[0].stake);
+        // and remove it from total participation
+        survey.participation = survey.participation.sub(previousParticipation);
+
+        // reset previously voted options, if any
+        delete survey.votes[msg.sender];
+    }
+
     /**
     * @notice Vote for multiple options in survey #`_surveyId`.
-    * @dev Empty options and stakes arrays allow to remove previous votes
     * @param _surveyId Id for survey
-    * @param _options Array with indexes of supported options
+    * @param _optionIds Array with indexes of supported options
     * @param _stakes Number of tokens assigned to each option
     */
-    function voteOptions(uint256 _surveyId, uint256[] _options, uint256[] _stakes) public {
-        require(_options.length == _stakes.length);
-        require(canVote(_surveyId, msg.sender));
+    function voteOptions(uint256 _surveyId, uint256[] _optionIds, uint256[] _stakes) public {
+        require(_optionIds.length == _stakes.length && _optionIds.length > 0);
 
         SurveyStruct storage survey = surveys[_surveyId];
 
-        uint256 voterStake = token.balanceOfAt(msg.sender, survey.snapshotBlock);
-        MultiOptionVote storage previouslyVoted = survey.voters[msg.sender];
-
-        // revert previous votes, if any
-        if (previouslyVoted.optionsCastedLength > 0) {
-            // voter removes their vote
-            for (uint256 i = 1; i <= previouslyVoted.optionsCastedLength; i++) {
-                survey.votes[previouslyVoted.castedVotes[i].optionId] = survey.votes[previouslyVoted.castedVotes[i].optionId].sub(previouslyVoted.castedVotes[i].stake);
-            }
-            // remove previous vote participation
-            survey.participation = survey.participation.sub(voterStake.sub(previouslyVoted.castedVotes[0].stake));
-
-            // reset previously voted options, if any
-            delete survey.voters[msg.sender];
-        }
+        // revert previous votes, if any (also checks if canVote)
```

Comment:
What do you think about making `canVote` a modifier and then using an internal `_resetVote()` function that both `voteOptions()` and `resetVote()` use?

I think it's better to make these requirements as explicit as possible.